### PR TITLE
Add dev version of spark to cross version tests 

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

Add the dev version of spark to the cross version tests config.

## How is this patch tested?

Existing unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/ml-package-versions.yml [hunk 1]
```diff
@@ -275,6 +275,16 @@ statsmodels:
 spark:
   package_info:
     pip_release: "pyspark"
+    install_dev: |
+      temp_dir=$(mktemp -d)
+      git clone --depth 1 https://github.com/apache/spark $temp_dir
+      cd $temp_dir
+      export MAVEN_OPTS="-Xmx2g -XX:ReservedCodeCacheSize=1g"
+      export MAVEN_CLI_OPTS="--no-transfer-progress"
+      ./build/mvn -DskipTests clean package
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:284]: The build roughly takes 25  ~ 30 min.

File: tests/pyfunc/test_spark.py [hunk 2]
```diff
@@ -56,11 +56,7 @@ def configure_environment():
 
 def get_spark_session(conf):
     conf.set(key="spark_session.python.worker.reuse", value=True)
-    return (
-        pyspark.sql.SparkSession.builder.config(conf=conf)
-        .master("local-cluster[2, 1, 1024]")
-        .getOrCreate()
-    )
+    return pyspark.sql.SparkSession.builder.config(conf=conf).master("local[*]").getOrCreate()
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:59]: `local-cluser` doesn't seem to work: https://github.com/mlflow/mlflow/runs/2388530643

File: tests/spark/test_spark_model_export.py [hunk 3]
```diff
@@ -58,11 +58,11 @@ def spark_custom_env(tmpdir):
...
-    conf.set(
-        key="spark.jars.packages",
-        value="ml.combust.mleap:mleap-spark-base_2.11:0.12.0,"
-        "ml.combust.mleap:mleap-spark_2.11:0.12.0",
-    )
+    # conf.set(
+    #     key="spark.jars.packages",
+    #     value="ml.combust.mleap:mleap-spark-base_2.11:0.12.0,"
+    #     "ml.combust.mleap:mleap-spark_2.11:0.12.0",
+    # )
```
Comment (harupy) [inline tests/spark/test_spark_model_export.py:65]: This seems to be the root cause of the failure: https://github.com/mlflow/mlflow/runs/2388530643

Comment (harupy) [reply tests/spark/test_spark_model_export.py:65]: Opened a PR to address this issue: https://github.com/mlflow/mlflow/pull/4268

File: mlflow/ml-package-versions.yml [hunk 4]
```diff
@@ -275,6 +275,15 @@ statsmodels:
...
     pip_release: "pyspark"
+    install_dev: |
+      temp_dir=$(mktemp -d)
+      git clone --depth 1 https://github.com/apache/spark $temp_dir
+      cd $temp_dir
+      export MAVEN_OPTS="-Xmx2g -XX:ReservedCodeCacheSize=1g"
+      ./build/mvn -DskipTests --no-transfer-progress clean package
+      cd python
+      python setup.py sdist
+      pip install dist/*.tar.gz
```
Comment (WeichenXu123) [inline mlflow/ml-package-versions.yml:286]: Is it possible to cache the build result ? and set cache expire time to be 1 day.
the compiling will took a long time.

Comment (harupy) [reply mlflow/ml-package-versions.yml:286]: I think it's possible. Let me try.

Comment (harupy) [reply mlflow/ml-package-versions.yml]: offline discussion: We'll address this in a follow-up.

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**